sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -232,15 +245,22 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
   private static class CalcFn extends DoFn<Row, Row> {
     private final String processElementBlock;
     private final Schema outputSchema;
+    private final List<String> jarPaths;
     private transient @Nullable ScriptEvaluator se = null;
 
-    public CalcFn(String processElementBlock, Schema outputSchema) {
+    public CalcFn(String processElementBlock, Schema outputSchema, List<String> jarPaths) {
       this.processElementBlock = processElementBlock;
       this.outputSchema = outputSchema;
+      this.jarPaths = jarPaths;
     }
 
-    ScriptEvaluator compile() {
+    ScriptEvaluator compile() throws IOException {
       ScriptEvaluator se = new ScriptEvaluator();
+      if (!jarPaths.isEmpty()) {
+        JavaUdfLoader udfLoader = new JavaUdfLoader();
+        ClassLoader classLoader = udfLoader.createClassLoader(jarPaths);
+        se.setParentClassLoader(classLoader);
```

Comment:
From the java doc I cannot really tell what this `setParentClassLoader` behave: https://janino-compiler.github.io/janino/apidocs/org/codehaus/janino/ScriptEvaluator.html#setParentClassLoader(java.lang.ClassLoader)


Does it use `classLoader` to replace the default one or combine with the default one? Basically I guess to make `ScriptEvaluator` work functionally, it needs to know both Beam and UDF implementation.

Reply:
The new classloader replaces the default one, but that is okay. The classloader we construct does not only contain the UDF jars. It also has a parent classloader, which is the same as the default one.